includes/sanitizers/class-amp-meta-sanitizer.php

```diff
@@ -0,0 +1,333 @@
+	 *
+	 * @var array $tags {
+	 *     An array of meta tag groupings.
+	 *
+	 *     @type DOMElement[] $charset        Charset meta tag(s).
+	 *     @type DOMElement[] $viewport       Viewport meta tag(s).
+	 *     @type DOMElement[] $amp_script_src <amp-script> source meta tags.
+	 *     @type DOMElement[] $other          Remaining meta tags.
+	 * }
+	 */
+	protected $meta_tags = [
+		self::TAG_CHARSET        => [],
+		self::TAG_VIEWPORT       => [],
+		self::TAG_AMP_SCRIPT_SRC => [],
+		self::TAG_OTHER          => [],
+	];
+
+	/**
+	 * Charset to use for AMP markup.
+	 *
+	 * @var string
+	 */
+	const AMP_CHARSET = 'utf-8';
+
+	/**
+	 * Viewport settings to use for AMP markup.
+	 *
+	 * @var string
+	 */
+	const AMP_VIEWPORT = 'width=device-width';
+
+	/**
+	 * Sanitize.
+	 */
+	public function sanitize() {
+		$this->xpath = new DOMXPath( $this->dom );
+		$this->head  = $this->ensure_head_is_present();
+		$charset     = $this->detect_charset();
+		$elements    = $this->dom->getElementsByTagName( static::$tag );
+
+		// Remove all nodes for easy reordering later on.
+		$elements = array_map(
+			static function ( $element ) {
+				return $element->parentNode->removeChild( $element );
+			},
+			iterator_to_array( $elements, false )
+		);
+
+		foreach ( $elements as $element ) {
+			/**
+			 * Meta tag to process.
+			 *
+			 * @var DOMElement $element
+			 */
+			if ( $element->hasAttribute( 'charset' ) ) {
+				$this->meta_tags[ self::TAG_CHARSET ][] = $element;
+			} elseif ( 'viewport' === $element->getAttribute( 'name' ) ) {
+				$this->meta_tags[ self::TAG_VIEWPORT ][] = $element;
+			} elseif ( 'amp-script-src' === $element->getAttribute( 'name' ) ) {
+				$this->meta_tags[ self::TAG_AMP_SCRIPT_SRC ][] = $element;
+			} else {
+				$this->meta_tags[ self::TAG_OTHER ][] = $element;
+			}
+		}
+
+		$this->ensure_charset_is_present( $charset );
+
+		if ( ! $this->is_correct_charset() ) { // phpcs:ignore Generic.CodeAnalysis.EmptyStatement
+			// @TODO Re-encode the content into UTF-8.
+			// ... sure?
+		}
+
+		$this->ensure_viewport_is_present();
+
+		$this->process_amp_script_meta_tags();
+
+		$this->re_add_meta_tags_in_optimized_order();
+	}
+
+	/**
+	 * Ensure that the <head> element is present in the document.
+	 *
+	 * @return DOMElement The document's <head> element.
+	 */
+	protected function ensure_head_is_present() {
+		$head = $this->dom->getElementsByTagName( 'head' )->item( 0 );
+
+		if ( ! $head ) {
+			$head = $this->dom->createElement( 'head' );
+			$head = $this->dom->documentElement->insertBefore( $head, $this->dom->documentElement->firstChild );
+		}
+
+		return $head;
+	}
+
+	/**
+	 * Detect the charset of the document.
+	 *
+	 * @return string|false Detected charset of the document, or false if none.
+	 */
+	protected function detect_charset() {
+		$charset = false;
+
+		// Check for HTML 4 http-equiv meta tags.
+		$http_equiv_tag = $this->xpath->query( '//meta[ @http-equiv and @content ]' )->item( 0 );
```

Comment:
```suggestion
		$http_equiv_tag = $this->xpath->query( '//meta[ translate(@http-equiv, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz') = 'content-type' and @content ]' )->item( 0 );
```

Alas, `lower-case()` is only in XPath 2.